src/objective/rank_obj.cu

```diff
@@ -483,6 +723,11 @@ class SortedLabelList {
       const GradientPair in_neg_gpair(-g * weight, 2.0f * weight * h);
       dh::AtomicAddGpair(&out_gpair[original_pos[neg_idx]], in_neg_gpair);
     });
+
+    // Wait until the computations done by the kernel is complete
+    dh::safe_cuda(cudaStreamSynchronize(nullptr));
+
+    weight_computer.ReleaseResources();
```

Comment:
`weight_computer.ReleaseResources();`

I am trying to move away from having objects in half configured states in xgboost. For example if you use an object in the wrong way or at the wrong time you would get a seg fault or some cryptic change of behaviour.

Instead we want to create an object on the stack when we need it, guarantee that it is in a usable state after construction, then let it be destroyed. This was previously hard to do without performance penalties because GPU memory allocation is expensive, but I think it is possible with caching allocators.

Reply:
please look. i have introduced an abstraction that lets the types be reference counted in fcb55e1. this will let the objects themselves keep track of how many copies they have.

Reply:
@sriramch That's impressive.  Never thought about adding a intrusive ptr in kernel.